mbtools/generate_quiz_csv.py

```diff
@@ -0,0 +1,140 @@
+import argparse
+from collections import defaultdict
+import os
+import csv
+from mbtools import utils
+from pathlib import Path
+
+
+def write_quiz_csvs(
+    quizzes_csv, quiz_questions_csv, quiz_multichoice_answers_csv, output_path
+):
+    with open(output_path / "quizzes.csv", "w") as outfile:
+        writer = csv.writer(outfile)
+        writer = csv.writer(outfile)
+        writer.writerow(quizzes_csv.keys())
+        writer.writerows(zip(*quizzes_csv.values()))
+
+    with open(output_path / "quiz_questions.csv", "w") as outfile:
+        writer = csv.writer(outfile)
+        writer = csv.writer(outfile)
+        writer.writerow(quiz_questions_csv.keys())
+        writer.writerows(zip(*quiz_questions_csv.values()))
+
+    with open(output_path / "quiz_multichoice_answers.csv", "w") as outfile:
+        writer = csv.writer(outfile)
+        writer = csv.writer(outfile)
+        writer.writerow(quiz_multichoice_answers_csv.keys())
+        writer.writerows(zip(*quiz_multichoice_answers_csv.values()))
+
+
+def order_questions(quiz):
+    # Organize Questions by Order
+    pages = defaultdict(list)
+    for question in quiz.quiz_questions:
+        pages[question.page].append(question)
+    for page in pages.values():
+        page.sort(key=lambda h: h.slot)
+    ordered_questions = []
+    ordered_keys = list(pages.keys())
+    ordered_keys.sort()
+    for key in ordered_keys:
+        ordered_questions.extend(pages[key])
+    return ordered_questions
+
+
+def generate_quiz_data(mbz_path, output_path):
+    question_bank = utils.parse_moodle_questionbank(mbz_path)
+    quizzes = utils.parse_backup_quizzes(mbz_path)
+
+    quizzes_csv = {
+        "quiz_name": [],
+        "question_number": [],
+        "question_id": []
+    }
+    quiz_questions_csv = {
+        "id": [],
+        "text": [],
+        "type": []
+    }
+    quiz_multichoice_answers_csv = {
+        "id": [],
+        "question_id": [],
+        "text": [],
+        "grade": [],
+        "feedback": []
+    }
+
+    answer_number = 0
+    for quiz in quizzes:
+        ordered_questions = order_questions(quiz)
+
+        question_number = 0
+
+        for moodle_question in ordered_questions:
+            q_bank_question = \
+                question_bank.get_question_by_entry(
+                    question_bank_entry_id=moodle_question.qbank_entry_id,
+                    version=moodle_question.version
+                )
+
+            id_number = q_bank_question.id_number
+            text = q_bank_question.etree.xpath('./questiontext')[0].text
+            question_type = q_bank_question.question_type
+
+            quizzes_csv["quiz_name"].append(quiz.name)
+            quizzes_csv["question_number"].append(question_number)
+            quizzes_csv["question_id"].append(id_number)
+
+            if id_number not in quiz_questions_csv["id"]:
```

Comment:
I'm wondering whether we should actually include this conditional / check. One one hand I can understand that it enforces an expected uniqueness, but it could also hide the fact that we accidentally injected duplicate IDs which we would just bite us later. Is there a functional reason for it being here for this script, or can we remove it so that we can use the output to help confirm all IDs are unique?

Reply:
Yes there is a reason, if there are questions that are getting used by multiple quizzes, it will eliminate them from showing up twice in the `quiz_questions.csv` file multiple times. They would otherwise show up because we loop through the quizzes to collect questions instead of parsing the question bank. 

If we were to get rid of the conditional, this would display the instances of the quiz questions... which is not necessarily wrong or bad, we'd just need to tell Katie and Iris.

Reply:
Ah, I see. That makes sense. I wonder if we can kill two birds with one stone, though, by gating this conditional on the tuple `(moodle_question.qbank_entry_id, moodle_question.version)` vs the `<idnumber>`. Doing this would give us a few things:

* To your original goal, it would de-duplicate entries where quizzes refer to the same question bank entry and version
* It would surface scenarios where our content might reference the same question bank entry but different versions (I think we always refer to the latest version of a question, but if not this would reflect the case that we have references to multiple versions of one question)
* If the scenario in the second bullet occurs, even when we inject UUIDs we'd end up with an unexpected duplicate because the `<idnumber>` is shared across versions (or that's my understanding). We'd have to deal with that implication if our content reflects that I guess, but I'm fairly certain it won't.

Thoughts?

Reply:
That sounds good to me. One question: In the case that this is triggered, what should we do? Print a warning out? Throw an error and cancel the job? I was thinking we could just print a warning.

Reply:
:thinking: -- Yeah, printing a warning sounds good :+1: .